Exclude platform-specific items using metadata

tmeschter opened the pull request:
### Description of Change

Currently the platform-specific files are included for every platform in the MSBuild evaluation, and then at build the  `_MauiRemovePlatformCompileItems` target removes _all_ these items and then adds back the ones for the current platform.

This works for an actual build but screws up VS, as the Project System doesn't properly handle a `Compile` item that is present in evaluation but removed during a build. The end result is that the platform-specific files are not actually treated as platform-specific, leading to a number of user-visible issues:

1. Platform-specific files incorrectly show all target frameworks in the Nav Bar
2. Spurious errors in the Error List
3. Squiggles in the editor, unless the user changes the Nav Bar to the "correct" target framework
4. Possibly they will not see actual errors in their code (e.g. if you use a Windows-only API in an Android-specific file, but the Nav Bar is set to Windows, you won't get an squiggle about that)
5. The spurious errors block Hot Reload as the Language Service thinks the project is very broken and will not attempt to apply changes

To work around this problem the Project System will now check items for special metadata indicating that they should be ignored. If an item is marked with `ExcludeFromCurrentConfiguration=true` the Project System will pretend it does not exist in the evaluation data for the current build configuration. And as the file no longer exists in the evaluation data, it no longer matters that the Project System does not properly handle the removal in the target.

Here we mark all items under the Platforms folder with `ExcludeFromCurrentConfiguration=true`, and then mark the items corresponding to the current platform with `ExcludeFromCurrentConfiguration=false`. The `_MauiRemovePlatformCompileItems` target is also updated to remove items based on the metadata rather than the folder structure.

### Issues Fixed

<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->

_Note that these issues also require a corresponding change in the [dotnet/project-system](https://github.com/dotnet/project-system) repo._

Fixes #4726.
Fixes #5294.

tmeschter committed:
Change handling of platform-specific files

Currently the platform-specific files are included for every platform in the MSBuild evaluation, and then at build the  _MauiRemovePlatformCompileItems target removes _all_ these items and then adds back the ones for the current platform.

This works for an actual build but screws up VS, as the Project System doesn't properly handle a `Compile` item that is present in evaluation but removed during a build. The end result is that the platform-specific files are not actually treated as platform specific, leading to a number of user-visible issues:

1. Platform-specific files incorrectly show all target frameworks in the Nav Bar.
2. Spurious errors in the Error List.
3. Squiggles in the editor, unless the user changes the Nav Bar to the "correct" target framework.
4. Possibly they will not see actual errors in their code (e.g. if you use a Windows-only API in an Android-specific file, but the Nav Bar is set to Windows, you won't get an squiggle about that).
5. The spurious errors block Hot Reload as the Language Service thinks the project is very broken and will not attempt to apply changes.

To work around this problem the Project System will now check items for special metadata indicating that they should be ignored. If an item is marked as a "LimitedConfigurationFile" but _not_ marked as "IncludeInThisConfiguration" the Project System will pretend it does not exist in the current build configuration. And as the file no longer exists in the evaluation data, it no longer matters that the Project System does not properly handle the removal in the target.

Here we mark all items under the Platforms folder with "LimitedConfigurationFile", and then mark the items corresponding to the current platform with "IncludeInThisConfiguration". The _MauiRemovePlatformCompileItems is updated to remove items based on the metadata rather than the folder structure.

tmeschter committed:
Simplify logic

Rather than two properties, we can use one. First all platform-specific files are tagged with `ExcludeFromCurrentConfiguration=true`. Then we update the platform-specific files for the current platform with `ExcludeFromCurrentConfiguration=false`, and the _MauiRemovePlatformCompileItems target can just check the one property.

jonathanpeppers commented on .nuspec/Microsoft.Maui.Controls.SingleProject.targets:
```diff
@@ -68,6 +68,44 @@
   </PropertyGroup>
 
+  <!-- Add metadata indicating that the platform-specific files are not part of every build configuration. -->
+  <ItemGroup Condition=" '$(SingleProject)' == 'true' ">
```
If each `<ItemGroup>` here checks `$(SingleProject)`, could we put all of them under a single `<ItemGroup>` element? Then the `Condition` would be on each `<Compile>` item.

tmeschter replied:
Done.

jonathanpeppers commented on .nuspec/Microsoft.Maui.Controls.SingleProject.targets:
```diff
@@ -68,6 +68,44 @@
+  </ItemGroup>
+
+  <!-- Add metadata for the files that are actually part of the current build configuration. -->
+  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))' == 'android' ">
```
Since this is an item group, it happens in a later MSBuild pass than properties. I think you could do this instead and it would work:
```suggestion
  <ItemGroup Condition=" '$(SingleProject)' == 'true' and '$(TargetPlatformIdentifier)' == 'android' ">
```

tmeschter replied:
Done.

Eilon commented:
@tmeschter - could this also solve https://github.com/dotnet/maui/issues/6642 ? Or does this seem different?

mattleibow commented:
/azp run

tmeschter commented:
> @tmeschter - could this also solve #6642 ? Or does this seem different?

@Eilon I haven't tried it out, but I think it will fix that bug as a side effect. Previously `_MauiRemovePlatformCompileItems` would remove all the files under "Platforms" and then add the platform-specific ones back, and that latter step pulled the Platforms\Android\test.cs file back in. Now the target only removes items.

tmeschter committed:
Simplify

1. Collapse the `ItemGroup`s into one, and move the conditions down to individual `Compile` items.
2. Since we're in an `ItemGroup` we can use `$(TargetPlatformIdentifier)` rather than `$([MSBuild]::GetTargetPlatformIdentifier($(TargetFramework)))`.

jonathanpeppers reviewed (approved):
Changes look OK.

@mattleibow @Redth do we need to manually test what happens if you use this inside VS with & without Tom's fixes in the Project System?

mattleibow commented:
/azp run

mattleibow commented:
I tested this in an old IDE:

> Microsoft Visual Studio Enterprise 2022 (64-bit) - Preview
> Version 17.2.0 Preview 3.0

And it all appears to be working as it previously was (still has all the TFMs in the dropdown):
 - moving files from one platform folder to another does not adjust the csproj
 - adding files to the platform folders does not adjust the csproj
 - building works for the csproj
 - intellisense works when selecting the correct tfm from the dropdown

Eilon commented:
@ThatGuyMike7 can you please file a new issue with references to the previous similar issues/PRs?
